pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -1568,15 +1617,33 @@ func (volTopology *wcpControllerVolumeTopology) GetTopologyInfoFromNodes(ctx con
 			var selectedSegments []map[string]string
 			for _, topology := range params.TopologyRequirement.GetPreferred() {
 				for label, value := range topology.GetSegments() {
-					clusterMoref, exists := azClusterMap[value]
-					if !exists || clusterMoref == "" {
-						return nil, logger.LogNewErrorf(log, "could not find the cluster MoID for zone %q in "+
-							"AvailabilityZone resources", value)
-					}
-					datastores, err := params.Vc.GetDatastoresByCluster(ctx, clusterMoref)
-					if err != nil {
-						return nil, logger.LogNewErrorf(log,
-							"failed to fetch datastores associated with cluster %q", clusterMoref)
+					var datastores []*cnsvsphere.DatastoreInfo
+					var err error
+					if isPodVMOnStretchedSupervisorEnabled {
+						clusterMorefs, exists := azClustersMap[value]
+						if !exists || len(clusterMorefs) == 0 {
+							return nil, logger.LogNewErrorf(log, "could not find the cluster MoIDs for zone %q in "+
+								"AvailabilityZone resources", value)
+						}
+						for _, clusterMoref := range clusterMorefs {
+							datastoresforClusterMoRef, err := params.Vc.GetDatastoresByCluster(ctx, clusterMoref)
```

Comment:
Suppose zone1 consists of Cluster1, Cluster2 and zone2 has just Cluster3.
Assume DS1 is accessible to Cluster2 and Cluster3.
With this logic, we will find node affinity as zone1, zone2 and error out, right?
Maybe we should check for shared datastores among the clusters of zone instead of combining all the datastores of the clusters?

Reply:
Do you mean fail if cross zonal datastore found?

Reply:
Yes - `selectedSegments` will have multiple AZ - zone1 and zon2 and we will fail volume provisioning with this code


```golang
			case numSelectedSegments > 1:
				// This situation will arise when datastore belongs to multiple zones but the
				// storageTopologyType is `zonal`. This seems like a configuration error.
				return nil, &common.InvalidTopologyProvisioningError{ErrMsg: fmt.Sprintf(
					"zonal volume is provisioned on %q datastore which is accessible from multiple zones: %+v. "+
						"Kindly check the configuration of the storage policy used in the StorageClass.",
					params.DatastoreURL, selectedSegments)}
```

Reply:
> Suppose zone1 consists of Cluster1, Cluster2 and zone2 has just Cluster3. Assume DS1 is accessible to Cluster2 and Cluster3. With this logic, we will find node affinity as zone1, zone2 and error out, right? Maybe we should check for shared datastores among the clusters of zone instead of combining all the datastores of the clusters?

IMO since GetTopologyInfoFromNodes() gets called to return node affinity in CreateVolume() response for already created volume, this situation will not arise since CreateVolume() will ensure that volume is confined to one of the zone only.